File: ts/test/entity/universal/UniversalEntity.test.ts

const envlocal = __dirname + '/../../../.env.local'
require('dotenv').config({ quiet: true, path: [envlocal] })

import { test, describe } from 'node:test'
import assert from 'node:assert'


import { UniversalManager, UniversalSDK, stdutil } from '../../..'

import {
  envOverride,
} from '../../utility'


describe('UniversalEntity', async () => {

  const um = new UniversalManager({ registry: __dirname + '/../../../test/registry' })
  const sdk = um.make('voxgig-solardemo')
  const entityMap: any = sdk._config.entity


  test('instance', async () => {
    const struct = sdk.utility().struct
    const items = struct.items

    items(entityMap, (item: any[]) => {
      const name = item[1].name
      const uent = sdk.Entity(name)
      assert(null != uent)
    })
  })


  const struct = sdk.utility().struct
  const items = struct.items

  items(entityMap, (item: any[]) => {
    const entityDef = item[1]
    const entityName = entityDef.name

    test('basic-' + entityName, async () => {
      const setup = await basicSetup(um, entityMap, entityName)
      const client = setup.client
      const struct = setup.struct

      const ops = entityDef.op || {}
      const ref = entityName + '_ref01'
      const ent = client.Entity(entityName)

      let createdData: any = null

      if (ops.create) {
        createdData = await testCreate(setup, ent, entityName, ref, entityDef)
      }

      if (ops.list) {
        await testList(setup, ent, entityDef, createdData, true)
      }

      if (ops.update && createdData) {
        await testUpdate(setup, ent, entityName, entityDef, createdData)
      }

      if (ops.load && createdData) {
        await testLoad(setup, ent, entityDef, createdData)
      }

      if (ops.remove && createdData) {
        await testRemove(setup, ent, entityDef, createdData)
      }

      if (ops.list && ops.remove && createdData) {
        await testList(setup, ent, entityDef, createdData, false)
      }
    })
  })

})



function resolveIdFields(data: any, idmap: any): any {
  const out: any = { ...data }
  for (const key of Object.keys(out)) {
    if (key.endsWith('_id')) {
      const baseRef = key.substring(0, key.length - 3) + '01'
      if (null != idmap[baseRef]) {
        out[key] = idmap[baseRef]
      }
    }
  }
  return out
}


async function testCreate(
  setup: any,
  ent: any,
  entityName: string,
  ref: string,
  entityDef: any,
) {
  let reqdata = resolveIdFields(setup.data.new[entityName][ref], setup.idmap)
  const resdata = await ent.create(reqdata)
  assert(null != resdata.id)
  return resdata
}


async function testList(
  setup: any,
  ent: any,
  entityDef: any,
  createdData: any,
  shouldExist: boolean,
) {
  const struct = setup.struct
  const isempty = struct.isempty
  const select = struct.select

  const matchFields = getDefaultTargetFields(entityDef, 'list')
  const match: any = {}
  for (const field of matchFields) {
    if (field !== 'id' && createdData && null != createdData[field]) {
      match[field] = createdData[field]
    }
  }

  const list = await ent.list(match)

  if (createdData) {
    if (shouldExist) {
      assert(!isempty(select(list, { id: createdData.id })))
    }
    else {
      assert(isempty(select(list, { id: createdData.id })))
    }
  }
}


async function testUpdate(
  setup: any,
  ent: any,
  entityName: string,
  entityDef: any,
  createdData: any,
) {
  const reqdata: any = {}
  reqdata.id = createdData.id

  const matchFields = getDefaultTargetFields(entityDef, 'update')
  for (const field of matchFields) {
    if (field !== 'id' && null != createdData[field]) {
      reqdata[field] = createdData[field]
    }
  }

  const textfield = findTextField(entityDef)
  let markdef: any = null

  if (textfield) {
    markdef = { name: textfield, value: 'Mark01-' + entityName + '_ref01_' + setup.now }
    reqdata[markdef.name] = markdef.value
  }

  const resdata = await ent.update(reqdata)
  assert(resdata.id === reqdata.id)

  if (markdef) {
    assert(resdata[markdef.name] === markdef.value)
  }
}


async function testLoad(
  setup: any,
  ent: any,
  entityDef: any,
  createdData: any,
) {
  const matchFields = getDefaultTargetFields(entityDef, 'load')
  const match: any = {}
  match.id = createdData.id
  for (const field of matchFields) {
    if (field !== 'id' && null != createdData[field]) {
      match[field] = createdData[field]
    }
  }

  const resdata = await ent.load(match)
  assert(resdata.id === createdData.id)
}


async function testRemove(
  setup: any,
  ent: any,
  entityDef: any,
  createdData: any,
) {
  const matchFields = getDefaultTargetFields(entityDef, 'remove')
  const match: any = {}
  match.id = createdData.id
  for (const field of matchFields) {
    if (field !== 'id' && null != createdData[field]) {
      match[field] = createdData[field]
    }
  }

  await ent.remove(match)
}



function getDefaultTargetFields(entityDef: any, opname: string): string[] {
  const op = entityDef.op?.[opname]
  if (!op) return []
  const targets = op.targets || []
  for (let i = targets.length - 1; i >= 0; i--) {
    if (!targets[i].select?.$action) {
      return targets[i].select?.exist || []
    }
  }
  return []
}


function findTextField(entityDef: any): string | null {
  for (const field of entityDef.fields || []) {
    if (field.type === '`$STRING`' && field.name !== 'id' && !field.name.endsWith('_id')) {
      return field.name
    }
  }
  return null
}


function makeEntityTestData(entityDef: any) {
  const fields = entityDef.fields || []
  const name = entityDef.name

  const data: any = {
    existing: { [name]: {} },
    new: { [name]: {} }
  }

  const idcount = 3
  const refs = Array.from({ length: idcount }, (_, i) =>
    `${name}${String(i).padStart(2, '0')}`)

  const idmapLocal = refs.reduce((a: any, ref) => (a[ref] = ref.toUpperCase(), a), {})

  let idx = 1
  for (const ref of refs) {
    const id = idmapLocal[ref]
    const ent: any = data.existing[name][id] = {}
    makeEntityTestFields(fields, idx++, ent)
    ent.id = id
  }

  const newRef = name + '_ref01'
  const newEnt: any = data.new[name][newRef] = {}
  makeEntityTestFields(fields, idx++, newEnt)
  delete newEnt.id

  return data
}


function makeEntityTestFields(fields: any[], start: number, entdata: any) {
  let num = start * fields.length * 10
  for (const field of fields) {
    entdata[field.name] =
      field.name.endsWith('_id') ?
        field.name.substring(0, field.name.length - 3).toUpperCase() + '01' :
        '`$NUMBER`' === field.type ? num :
          '`$BOOLEAN`' === field.type ? 0 === num % 2 :
            '`$OBJECT`' === field.type ? {} :
              '`$MAP`' === field.type ? {} :
                '`$ARRAY`' === field.type ? [] :
                  '`$LIST`' === field.type ? [] :
                    's' + (num.toString(16))
    num++
  }
}


async function basicSetup(um: any, entityMap: any, entityName: string, extra?: any) {
  const options: any = {}

  const allExisting: any = {}
  const allNew: any = {}

  const struct = stdutil.struct
  const items = struct.items
  const flatten = struct.flatten

  items(entityMap, (item: any[]) => {
    const entityDef = item[1]
    const testData = makeEntityTestData(entityDef)
    Object.assign(allExisting, testData.existing)
    Object.assign(allNew, testData.new)
  })

  options.entity = allExisting

  const sdk = um.make('voxgig-solardemo')
  let client = sdk.test(options, { ref: 'voxgig-solardemo', model: sdk._options.model })

  const clientStruct = client.utility().struct
  const merge = clientStruct.merge
  const transform = clientStruct.transform

  const idEntries: string[] = []
  items(entityMap, (item: any[]) => {
    const ename = item[1].name
    for (let i = 0; i < 3; i++) {
      idEntries.push(`${ename}${String(i).padStart(2, '0')}`)
    }
  })

  let idmap = transform(
    idEntries,
    {
      '`$PACK`': ['', {
        '`$KEY`': '`$COPY`',
        '`$VAL`': ['`$FORMAT`', 'upper', '`$COPY`']
      }]
    })

  const env = envOverride({
    'UNIVERSAL_TEST_ENTID': idmap,
    'UNIVERSAL_TEST_LIVE': 'FALSE',
    'UNIVERSAL_TEST_EXPLAIN': 'FALSE',
    'UNIVERSAL_APIKEY': 'NONE',
  })

  idmap = env['UNIVERSAL_TEST_ENTID']

  if ('TRUE' === env.UNIVERSAL_TEST_LIVE) {
    const liveopts: any = {
      ref: 'voxgig-solardemo',
      model: sdk._options.model,
      apikey: env.UNIVERSAL_APIKEY,
    }
    client = new UniversalSDK(um, null != extra ? merge([liveopts, extra]) : liveopts)

    // Discover real parent entity IDs from the live API.
    for (const item of Object.values(entityMap) as any[]) {
      const eDef = item
      const eName = eDef.name
      const listOp = eDef.op?.list
      const listTarget = listOp?.targets?.[0]
      if (null == listTarget) continue

      const listParams = listTarget.args?.params || []
      if (listParams.length > 0) continue // skip nested entities

      const listPath = (listTarget.parts || []).join('/')
      const res: any = await client.direct({ path: listPath, method: 'GET', params: {} })
      if (res.ok && Array.isArray(res.data)) {
        for (let i = 0; i < Math.min(res.data.length, 3); i++) {
          const ref = `${eName}${String(i).padStart(2, '0')}`
          idmap[ref] = res.data[i].id
        }
      }
    }
  }

  const setup = {
    idmap,
    env,
    options,
    client,
    struct: client.utility().struct,
    data: { existing: allExisting, new: allNew },
    explain: 'TRUE' === env.UNIVERSAL_TEST_EXPLAIN,
    now: Date.now(),
  }

  return setup
}

File: go/test/entity/universal/universal_entity_test.go
package universal

import (
	"fmt"
	"os"
	"strconv"
	"strings"
	"testing"
	"time"

	sdk "voxgiguniversalsdk"
	"voxgiguniversalsdk/core"

	vs "github.com/voxgig/struct"
)

func TestUniversalEntity(t *testing.T) {
	um := sdk.NewUniversalManager(map[string]any{
		"registry": "../../../test/registry",
	})
	baseSDK := um.Make("voxgig-solardemo")
	entityMap := core.ToMapAny(vs.GetPath([]any{"entity"}, baseSDK.GetRootCtx().Config))

	t.Run("instance", func(t *testing.T) {
		for _, item := range vs.Items(entityMap) {
			entDef := core.ToMapAny(item[1])
			name, _ := vs.GetProp(entDef, "name").(string)
			uent := baseSDK.Entity(name, nil)
			if uent == nil {
				t.Errorf("Entity(%q) should not be nil", name)
			}
		}
	})

	for _, item := range vs.Items(entityMap) {
		entDef := core.ToMapAny(item[1])
		entityName, _ := vs.GetProp(entDef, "name").(string)

		t.Run("basic-"+entityName, func(t *testing.T) {
			setup := basicSetup(t, um, entityMap, entityName)
			client := setup.client

			ops := core.ToMapAny(vs.GetProp(entDef, "op"))
			if ops == nil {
				ops = map[string]any{}
			}
			ref := entityName + "_ref01"
			ent := client.Entity(entityName, nil)

			var createdData map[string]any

			if vs.GetProp(ops, "create") != nil {
				createdData = testCreate(t, setup, ent, entityName, ref, entDef)
			}

			if vs.GetProp(ops, "list") != nil {
				testList(t, setup, ent, entDef, createdData, true)
			}

			if vs.GetProp(ops, "update") != nil && createdData != nil {
				testUpdate(t, setup, ent, entityName, entDef, createdData)
			}

			if vs.GetProp(ops, "load") != nil && createdData != nil {
				testLoad(t, setup, ent, entDef, createdData)
			}

			if vs.GetProp(ops, "remove") != nil && createdData != nil {
				testRemove(t, setup, ent, entDef, createdData)
			}

			if vs.GetProp(ops, "list") != nil && vs.GetProp(ops, "remove") != nil && createdData != nil {
				testList(t, setup, ent, entDef, createdData, false)
			}
		})
	}
}

func resolveIdFields(data map[string]any, idmap map[string]any) map[string]any {
	out := map[string]any{}
	for k, v := range data {
		out[k] = v
	}
	for key := range out {
		if strings.HasSuffix(key, "_id") {
			baseRef := key[:len(key)-3] + "01"
			if idmap[baseRef] != nil {
				out[key] = idmap[baseRef]
			}
		}
	}
	return out
}

func testCreate(
	t *testing.T,
	setup *entityTestSetup,
	ent sdk.UniversalEntity,
	entityName string,
	ref string,
	entityDef map[string]any,
) map[string]any {
	t.Helper()

	newData := core.ToMapAny(vs.GetPath([]any{"new", entityName, ref}, setup.data))
	if newData == nil {
		t.Fatalf("No new data for %s/%s", entityName, ref)
	}
	reqdata := resolveIdFields(newData, setup.idmap)

	result, err := ent.Create(reqdata, nil)
	if err != nil {
		t.Fatalf("Create failed: %v", err)
	}
	resdata := core.ToMapAny(result)
	if resdata == nil {
		t.Fatal("Create result should be a map")
	}
	if resdata["id"] == nil {
		t.Fatal("Created entity should have an id")
	}
	return resdata
}

func testList(
	t *testing.T,
	setup *entityTestSetup,
	ent sdk.UniversalEntity,
	entityDef map[string]any,
	createdData map[string]any,
	shouldExist bool,
) {
	t.Helper()

	matchFields := getDefaultTargetFields(entityDef, "list")
	match := map[string]any{}
	for _, field := range matchFields {
		if field != "id" && createdData != nil && createdData[field] != nil {
			match[field] = createdData[field]
		}
	}

	result, err := ent.List(match, nil)
	if err != nil {
		t.Fatalf("List failed: %v", err)
	}

	if createdData != nil {
		list := toSlice(result)
		found := vs.Select(list, map[string]any{"id": createdData["id"]})

		if shouldExist {
			if vs.IsEmpty(found) {
				t.Fatalf("Expected created entity %v in list", createdData["id"])
			}
		} else {
			if !vs.IsEmpty(found) {
				t.Fatalf("Expected created entity %v NOT in list after remove", createdData["id"])
			}
		}
	}
}

func testUpdate(
	t *testing.T,
	setup *entityTestSetup,
	ent sdk.UniversalEntity,
	entityName string,
	entityDef map[string]any,
	createdData map[string]any,
) {
	t.Helper()

	reqdata := map[string]any{}
	reqdata["id"] = createdData["id"]

	matchFields := getDefaultTargetFields(entityDef, "update")
	for _, field := range matchFields {
		if field != "id" && createdData[field] != nil {
			reqdata[field] = createdData[field]
		}
	}

	textfield := findTextField(entityDef)
	var markName, markValue string

	if textfield != "" {
		markName = textfield
		markValue = "Mark01-" + entityName + "_ref01_" + strconv.FormatInt(setup.now, 10)
		reqdata[markName] = markValue
	}

	result, err := ent.Update(reqdata, nil)
	if err != nil {
		t.Fatalf("Update failed: %v", err)
	}
	resdata := core.ToMapAny(result)
	if resdata == nil {
		t.Fatal("Update result should be a map")
	}
	if resdata["id"] != reqdata["id"] {
		t.Fatalf("Expected id=%v, got %v", reqdata["id"], resdata["id"])
	}

	if markName != "" {
		if resdata[markName] != markValue {
			t.Fatalf("Expected %s=%v, got %v", markName, markValue, resdata[markName])
		}
	}
}

func testLoad(
	t *testing.T,
	setup *entityTestSetup,
	ent sdk.UniversalEntity,
	entityDef map[string]any,
	createdData map[string]any,
) {
	t.Helper()

	matchFields := getDefaultTargetFields(entityDef, "load")
	match := map[string]any{}
	match["id"] = createdData["id"]
	for _, field := range matchFields {
		if field != "id" && createdData[field] != nil {
			match[field] = createdData[field]
		}
	}

	result, err := ent.Load(match, nil)
	if err != nil {
		t.Fatalf("Load failed: %v", err)
	}
	resdata := core.ToMapAny(result)
	if resdata == nil {
		t.Fatal("Load result should be a map")
	}
	if resdata["id"] != createdData["id"] {
		t.Fatalf("Expected id=%v, got %v", createdData["id"], resdata["id"])
	}
}

func testRemove(
	t *testing.T,
	setup *entityTestSetup,
	ent sdk.UniversalEntity,
	entityDef map[string]any,
	createdData map[string]any,
) {
	t.Helper()

	matchFields := getDefaultTargetFields(entityDef, "remove")
	match := map[string]any{}
	match["id"] = createdData["id"]
	for _, field := range matchFields {
		if field != "id" && createdData[field] != nil {
			match[field] = createdData[field]
		}
	}

	_, err := ent.Remove(match, nil)
	if err != nil {
		t.Fatalf("Remove failed: %v", err)
	}
}

func getDefaultTargetFields(entityDef map[string]any, opname string) []string {
	op := core.ToMapAny(vs.GetPath([]any{"op", opname}, entityDef))
	if op == nil {
		return []string{}
	}
	targets, _ := vs.GetProp(op, "targets").([]any)
	if targets == nil {
		return []string{}
	}
	for i := len(targets) - 1; i >= 0; i-- {
		tm := core.ToMapAny(targets[i])
		if tm == nil {
			continue
		}
		sel := core.ToMapAny(vs.GetProp(tm, "select"))
		if sel != nil && vs.GetProp(sel, "$action") != nil {
			continue
		}
		exist, _ := vs.GetProp(sel, "exist").([]any)
		if exist == nil {
			return []string{}
		}
		out := make([]string, 0, len(exist))
		for _, e := range exist {
			if s, ok := e.(string); ok {
				out = append(out, s)
			}
		}
		return out
	}
	return []string{}
}

func findTextField(entityDef map[string]any) string {
	fields, _ := vs.GetProp(entityDef, "fields").([]any)
	for _, f := range fields {
		fm := core.ToMapAny(f)
		if fm == nil {
			continue
		}
		ftype, _ := vs.GetProp(fm, "type").(string)
		fname, _ := vs.GetProp(fm, "name").(string)
		if ftype == "`$STRING`" && fname != "id" && !strings.HasSuffix(fname, "_id") {
			return fname
		}
	}
	return ""
}

func makeEntityTestData(entityDef map[string]any) map[string]any {
	fields, _ := vs.GetProp(entityDef, "fields").([]any)
	if fields == nil {
		fields = []any{}
	}
	name, _ := vs.GetProp(entityDef, "name").(string)

	existing := map[string]any{}
	newData := map[string]any{}

	idcount := 3
	refs := make([]string, idcount)
	idmapLocal := map[string]string{}
	for i := 0; i < idcount; i++ {
		ref := fmt.Sprintf("%s%02d", name, i)
		refs[i] = ref
		idmapLocal[ref] = strings.ToUpper(ref)
	}

	idx := 1
	for _, ref := range refs {
		id := idmapLocal[ref]
		ent := map[string]any{}
		makeEntityTestFields(fields, idx, ent)
		idx++
		ent["id"] = id
		existing[id] = ent
	}

	newRef := name + "_ref01"
	newEnt := map[string]any{}
	makeEntityTestFields(fields, idx, newEnt)
	delete(newEnt, "id")
	newData[newRef] = newEnt

	return map[string]any{
		"existing": map[string]any{name: existing},
		"new":      map[string]any{name: newData},
	}
}

func makeEntityTestFields(fields []any, start int, entdata map[string]any) {
	num := start * len(fields) * 10
	for _, f := range fields {
		fm := core.ToMapAny(f)
		if fm == nil {
			continue
		}
		fname, _ := vs.GetProp(fm, "name").(string)
		ftype, _ := vs.GetProp(fm, "type").(string)

		if strings.HasSuffix(fname, "_id") {
			entdata[fname] = strings.ToUpper(fname[:len(fname)-3]) + "01"
		} else if ftype == "`$NUMBER`" {
			entdata[fname] = num
		} else if ftype == "`$BOOLEAN`" {
			entdata[fname] = num%2 == 0
		} else if ftype == "`$OBJECT`" || ftype == "`$MAP`" {
			entdata[fname] = map[string]any{}
		} else if ftype == "`$ARRAY`" || ftype == "`$LIST`" {
			entdata[fname] = []any{}
		} else {
			entdata[fname] = "s" + strconv.FormatInt(int64(num), 16)
		}
		num++
	}
}

type entityTestSetup struct {
	idmap   map[string]any
	env     map[string]string
	client  *sdk.UniversalSDK
	data    map[string]any
	explain bool
	now     int64
}

func basicSetup(
	t *testing.T,
	um *sdk.UniversalManager,
	entityMap map[string]any,
	entityName string,
) *entityTestSetup {
	t.Helper()

	allExisting := map[string]any{}
	allNew := map[string]any{}

	for _, item := range vs.Items(entityMap) {
		entDef := core.ToMapAny(item[1])
		testData := makeEntityTestData(entDef)

		existMap := core.ToMapAny(vs.GetProp(testData, "existing"))
		for k, v := range existMap {
			allExisting[k] = v
		}
		newMap := core.ToMapAny(vs.GetProp(testData, "new"))
		for k, v := range newMap {
			allNew[k] = v
		}
	}

	options := map[string]any{
		"entity": allExisting,
	}

	sdkInst := um.Make("voxgig-solardemo")
	client := sdkInst.Test(options, nil)

	// Build idmap
	idEntries := []any{}
	for _, item := range vs.Items(entityMap) {
		entDef := core.ToMapAny(item[1])
		ename, _ := vs.GetProp(entDef, "name").(string)
		for i := 0; i < 3; i++ {
			idEntries = append(idEntries, fmt.Sprintf("%s%02d", ename, i))
		}
	}

	idmap := core.ToMapAny(vs.Transform(
		idEntries,
		map[string]any{
			"`$PACK`": []any{"", map[string]any{
				"`$KEY`": "`$COPY`",
				"`$VAL`": []any{"`$FORMAT`", "upper", "`$COPY`"},
			}},
		},
	))
	if idmap == nil {
		idmap = map[string]any{}
	}

	env := envOverride(map[string]string{
		"UNIVERSAL_TEST_LIVE":    "FALSE",
		"UNIVERSAL_TEST_EXPLAIN": "FALSE",
		"UNIVERSAL_APIKEY":       "NONE",
	})

	if env["UNIVERSAL_TEST_LIVE"] == "TRUE" {
		liveopts := map[string]any{
			"ref":    "voxgig-solardemo",
			"model":  um.ResolveModel("voxgig-solardemo"),
			"apikey": env["UNIVERSAL_APIKEY"],
		}
		client = sdk.NewUniversalSDK(um, liveopts)

		// Discover real parent entity IDs from the live API.
		for _, item := range vs.Items(entityMap) {
			eDef := core.ToMapAny(item[1])
			eName, _ := vs.GetProp(eDef, "name").(string)
			listOp := core.ToMapAny(vs.GetPath([]any{"op", "list"}, eDef))
			if listOp == nil {
				continue
			}
			targets, _ := vs.GetProp(listOp, "targets").([]any)
			if len(targets) == 0 {
				continue
			}
			listTarget := core.ToMapAny(targets[0])
			if listTarget == nil {
				continue
			}
			listParams, _ := vs.GetPath([]any{"args", "params"}, listTarget).([]any)
			if len(listParams) > 0 {
				continue // skip nested entities
			}
			parts, _ := vs.GetProp(listTarget, "parts").([]any)
			listPath := vs.Join(parts, "/", true)

			res, err := client.Direct(map[string]any{
				"path": listPath, "method": "GET", "params": map[string]any{},
			})
			if err == nil && res["ok"] == true {
				if dataList, ok := res["data"].([]any); ok {
					for i := 0; i < len(dataList) && i < 3; i++ {
						if dm := core.ToMapAny(dataList[i]); dm != nil {
							ref := fmt.Sprintf("%s%02d", eName, i)
							idmap[ref] = dm["id"]
						}
					}
				}
			}
		}
	}

	return &entityTestSetup{
		idmap:   idmap,
		env:     env,
		client:  client,
		data:    map[string]any{"existing": allExisting, "new": allNew},
		explain: env["UNIVERSAL_TEST_EXPLAIN"] == "TRUE",
		now:     time.Now().UnixMilli(),
	}
}

func envOverride(m map[string]string) map[string]string {
	live := os.Getenv("UNIVERSAL_TEST_LIVE")
	override := os.Getenv("UNIVERSAL_TEST_OVERRIDE")

	if live == "TRUE" || override == "TRUE" {
		for k := range m {
			if envval := os.Getenv(k); envval != "" {
				m[k] = strings.TrimSpace(envval)
			}
		}
	}

	if explain := os.Getenv("UNIVERSAL_TEST_EXPLAIN"); explain != "" {
		m["UNIVERSAL_TEST_EXPLAIN"] = explain
	}

	return m
}

// toSlice converts the result from List (which may be []any or a
// wrapped entity list) into a plain []any of maps.
func toSlice(val any) []any {
	if val == nil {
		return []any{}
	}
	if list, ok := val.([]any); ok {
		out := make([]any, 0, len(list))
		for _, item := range list {
			if ent, ok := item.(sdk.Entity); ok {
				out = append(out, ent.Data())
			} else {
				out = append(out, item)
			}
		}
		return out
	}
	return []any{}
}
